external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/MaxUncommittedOffsetTest.java

```diff
@@ -0,0 +1,245 @@
+        Time.advanceTime(commitOffsetPeriodMs + KafkaSpout.TIMER_DELAY_MS);
+        spout.nextTuple();
+    }
+
+    @Test
+    public void testNextTupleWillNotEmitMoreThanMaxUncommittedOffsetsPlusMaxPollRecordsMessages() {
+        //The upper bound on uncommitted offsets should be maxUncommittedOffsets + maxPollRecords
+        //This is reachable by emitting maxUncommittedOffsets messages, acking the first message, then polling.
+        try (Time.SimulatedTime simulatedTime = new Time.SimulatedTime()) {
+            //First check that maxUncommittedOffsets is respected when emitting from scratch
+            ArgumentCaptor<KafkaSpoutMessageId> messageIds = emitMaxUncommittedOffsetsMessagesAndCheckNoMoreAreEmitted(numMessages);
+            reset(collector);
+
+            failAllExceptTheFirstMessageThenCommit(messageIds);
+
+            //Offset 0 is acked, 1 to maxUncommittedOffsets - 1 are failed
+            //The spout should now emit another maxPollRecords messages
+            //This is allowed because the acked message brings the numUncommittedOffsets below the cap
+            for (int i = 0; i < maxUncommittedOffsets; i++) {
+                spout.nextTuple();
+            }
+
+            ArgumentCaptor<KafkaSpoutMessageId> secondRunMessageIds = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
+            verify(collector, times(maxPollRecords)).emit(
+                anyObject(),
+                anyObject(),
+                secondRunMessageIds.capture());
+            reset(collector);
+
+            List<Long> firstRunOffsets = messageIds.getAllValues().stream()
+                .map(messageId -> messageId.offset())
+                .collect(Collectors.toList());
+            List<Long> secondRunOffsets = secondRunMessageIds.getAllValues().stream()
+                .map(messageId -> messageId.offset())
+                .collect(Collectors.toList());
+            assertThat("Expected the newly emitted messages to have no overlap with the first batch", secondRunOffsets.removeAll(firstRunOffsets), is(false));
+
+            //Offset 0 is acked, 1 to maxUncommittedOffsets-1 are failed, maxUncommittedOffsets to maxUncommittedOffsets + maxPollRecords-1 are emitted
+            //There are now maxUncommittedOffsets-1 + maxPollRecords records emitted past the last committed offset
+            //Retry the failed messages and ack the first one.
+            Time.advanceTimeSecs(initialRetryDelaySecs);
+            for (int i = 0; i < numMessages; i++) {
+                spout.nextTuple();
+            }
+            ArgumentCaptor<KafkaSpoutMessageId> thirdRunMessageIds = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
+            verify(collector, times(maxPollRecords-1)).emit(
+                anyObject(),
+                anyObject(),
+                thirdRunMessageIds.capture());
+            reset(collector);
+
+            failAllExceptTheFirstMessageThenCommit(thirdRunMessageIds);
```

Comment:
Minor: might be better to add comparison between captured message ids and first run message ids but excluding first one to make sure failed tuples are reemitted.